### ISSUE TITLE
Copy fails when source file is identical to the destination and destination is a hard link

### ISSUE THREAD

**Issue by biggsean** [ansible/ansible-modules-core#302]
Issue Type:  Bug Report
Ansible Version:  ansible 1.7.2
Environment:  RHEL 6.4 running ansible - managing RHEL 6+ and Solaris 10+ systems
Summary:  Copy fails when source file is identical to the destination and destination is a hard link (links > 1).

Steps To Reproduce:
1) try to copy a file that is identical to a system where the destination is a hard link
2) get failure message (see below)

Expected Results:
I would expect that it would report no change if the files were identical, provided any supplied file attributes were also identical.

Actual Results:
    ### /etc/hosts has > 1 links
    [root@padishah1 ansible]# ls -l /etc/hosts  
    -rw-r--r--. 2 root root 13888 May 16 17:21 /etc/hosts
    [root@padishah1 ansible]# find /etc -xdev -samefile /etc/hosts
    /etc/sysconfig/networking/profiles/default/hosts
    /etc/hosts

```
### hosts.bkp = hosts && hosts.different != hosts
[root@padishah1 ansible]# md5sum /etc/hosts /etc/hosts.{bkp,different}
0f7cc9afcb0f17a9864b45391027a6ad  /etc/hosts
0f7cc9afcb0f17a9864b45391027a6ad  /etc/hosts.bkp
9c5799def96f6f2e7309a3914075589b  /etc/hosts.different

### Trying to use the copy module with an identical file fails
[root@padishah1 ansible]# ansible all -i <(echo -e "localhost ansible_connection=local") -m copy -a "src=/etc/hosts.bkp dest=/etc/hosts mode=0644 owner=root group=root" --check
localhost | FAILED >> {
    "failed": true,
    "md5sum": "0f7cc9afcb0f17a9864b45391027a6ad",
    "msg": "absolute paths are required"
}

### Trying to copy with a new file is sucessful
[root@padishah1 ansible]# ansible all -i <(echo -e "localhost ansible_connection=local") -m copy -a "src=/etc/hosts.different dest=/etc/hosts mode=0644 owner=root group=root" --check
--- before: /etc/hosts
+++ after: /etc/hosts.different
---diff redacted---
+
localhost | success >> {
    "changed": true
}

### This is not the case when links = 1
[root@padishah1 ansible]# cp -p --remove-destination /etc/hosts.bkp /etc/hosts
cp: overwrite `/etc/hosts'? yes
[root@padishah1 ansible]# ls -l /etc/hosts
-rw-r--r--. 1 root root 13888 May 16 17:21 /etc/hosts

### Identical file is now successful (changed: false is expected)
[root@padishah1 ansible]# ansible all -i <(echo -e "localhost ansible_connection=local") -m copy -a "src=/etc/hosts.bkp dest=/etc/hosts mode=0644 owner=root group=root" --check

localhost | success >> {
   "changed": false,
    "dest": "/etc/hosts",
    "gid": 0,
    "group": "root",
    "md5sum": "0f7cc9afcb0f17a9864b45391027a6ad",
    "mode": "0644",
    "owner": "root",
    "path": "/etc/hosts",
    "secontext": "unconfined_u:object_r:etc_t:s0",
    "size": 13888,
    "state": "file",
    "uid": 0
}

[root@padishah1 ansible]# ansible all -i <(echo -e "localhost ansible_connection=local") -m copy -a "src=/etc/hosts.different dest=/etc/hosts mode=0644 owner=root group=root" --check
--- before: /etc/hosts
+++ after: /etc/hosts.different
---diff redacted---
+

localhost | success >> {
    "changed": true
}
```

**Comment by sysadmin75** [comment]
My old PR was done on the devel branch.

This new PR was done using a new branch.
https://github.com/ansible/ansible/pull/9806

**Comment by bcoca** [comment]
closing in favor of PR
